### DISCUSSION TITLE
Auto-submitting the form on blur?

### DISCUSSION THREAD

**Discussion by piotr-sobczyk** [react-hook-form/react-hook-form#2494] [Q&A] [answered]
We have a requirement, that there should be no submit button on our form but instead it should auto-submit on every blur and send the data to the server, but only if all the fields are passing the validation.

Currently we do it this way:
```
<form>
  <input name="field1" ref={register({ required: true })} onBlur={formChanged} />
  <input name="field2" ref={register({ required: true })} onBlur={formChanged} />
  ... //etc
</form>
```

And our form config and `formChanged` handler:
```
const { register, formState, getValues } = useForm<CheckoutAddressData>({
    mode: 'onBlur',
});

const formChanged = () => {
  const { isValid } = formState;
  const formValues = getValues();
  if(isValid) {
     sendValuesToServer(formValues);
  }
}
```

However this doesn't as expected. When `formChanged` is triggered after the last field is filled, the `isValid` is still `false`. Only after the next blur cycle, the `isValid` has the `true` value. I guess the problem is that our `onBlur` handler runs before library's validation check.

Here's a sandbox demonstrating my problem:  https://codesandbox.io/s/musing-dhawan-f13kn .

Am we doing something very non-standard here? Is there a better pattern to get what we want, without the outdated validation issue and without manually listening to `onBlur` on every form field? Maybe there is some built-in mechanism in the library that I could use instead?

**Comment by bluebill1049** [comment]
can you share a simple codesandbox?

**Reply by piotr-sobczyk** [reply]
Sure, here you got: https://codesandbox.io/s/musing-dhawan-f13kn?file=/src/App.js . Please note that you need to blur one extra time to get the expected result.

**Reply by bluebill1049** [reply]
what about this? https://codesandbox.io/s/stupefied-dirac-b20ln?file=/src/App.js

**Reply by piotr-sobczyk** [reply]
It looks and works much better! But new it doesn't send new values to the server after the whole form was filled first time (and the user updates the form with another valid set of values).

**Reply by pprob** [reply]
Did you manage to find a solution in the end? I'm trying to achieve the same behaviour with my form. I've had a play around with bluebill's sandbox and I've tried using reset inside the conditional inside the useEffect to reset the form state & isDirty. However that triggers the useEffect whenever I change an input value because of the dependency on isDirty.
edit: debouncing the network request until a period of no user input seems to work ok for now.

**Comment by squiaios** [comment, marked as answer]
It would be more interesting to pass an onSubmit function to the formProvider and that this function be executed after the blur event of an input is triggerd (only if the validation is valid), or am I totally wrong?

Somethig like this :

```
const useFormProps = useForm({
        defaultValues: {url: ''},
        mode: 'onBlur',
        onSubmit: () => {
            console.log('function triggered after blur and validation valid')
        },
        resolver: joiResolver(Joi.object({
            url: Joi.string().uri({scheme: ['https']}),
        })),
    });

```

**Reply by bluebill1049** [reply]
can you try to attach onBlur at the form and then subscribe to isValid state?

**Reply by squiaios** [reply]
Thank you very much, it's working.

Here an exampl :

[![Edit interesting-chaum-bw5wq](https://codesandbox.io/static/img/play-codesandbox.svg)](https://codesandbox.io/s/interesting-chaum-bw5wq?fontsize=14&hidenavigation=1&theme=dark)

**Reply by itsuness** [reply]
Hey everyone I'm trying to do the same here but it should submit the value on each field blur

**Reply by Moshyfawn** [reply]
> Hey everyone I'm trying to do the same here but it should submit the value on each field blur

[The codesandbox](https://codesandbox.io/s/interesting-chaum-bw5wq?fontsize=14&hidenavigation=1&theme=dark) in [the above comment](https://github.com/react-hook-form/react-hook-form/discussions/2494#discussioncomment-1867033) does exactly that

**Reply by nathanmlui** [reply]
What if we only want to submit the field that experienced the onBlur event, not the entire form?

**Reply by carlos-lopez-SaRa** [reply]
@nathanmlui probably going to have to make 1 form wrapper per field...
